src/formula/variant.cpp

```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
-		delete list_;
-		break;
-	case TYPE_STRING:
-		delete string_;
-		break;
-	case TYPE_MAP:
-		delete map_;
-		break;
-	// These are not used here, add them to silence a compiler warning.
-	case TYPE_CALLABLE:
-	case TYPE_NULL:
-	case TYPE_DECIMAL:
-	case TYPE_INT :
-		break;
-	}
-}
 
-std::string variant::type_string() const {
-	return variant_type_to_string(type_);
-}
-
-variant::variant() : type_(TYPE_NULL), int_value_(0)
-{}
-
-variant::variant(int n) : type_(TYPE_INT), int_value_(n)
+variant::variant()
+	: value_()
 {}
 
-variant::variant(int n, variant::DECIMAL_VARIANT_TYPE /*type*/) : type_(TYPE_DECIMAL), decimal_value_(n)
-{}
-
-variant::variant(double n, variant::DECIMAL_VARIANT_TYPE /*type*/) : type_(TYPE_DECIMAL) {
-	n *= 1000;
-	decimal_value_ = static_cast<int>(n);
+variant::variant(int n)
+	: value_(std::make_shared<game_logic::variant_int>(n))
+{
+	assert(value_.get());
+}
 
-	n -= decimal_value_;
+variant::variant(int n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
+}
 
-	if(n > 0.5)
-		decimal_value_++;
-	else if(n < -0.5)
-		decimal_value_--;
+variant::variant(double n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
 }
 
 variant::variant(const game_logic::formula_callable* callable)
-	: type_(TYPE_CALLABLE), callable_(callable)
+	: value_(std::make_shared<game_logic::variant_callable>(callable))
 {
-	assert(callable_);
+	assert(value_.get());
 }
 
-variant::variant(std::vector<variant>* array)
-    : type_(TYPE_LIST)
+variant::variant(std::vector<variant>* vec)
+    : value_((std::make_shared<game_logic::variant_list>(*vec)))
 {
-	assert(array);
-	list_ = new std::vector<variant>(*array);
+	assert(value_.get());
 }
 
 variant::variant(const std::string& str)
-	: type_(TYPE_STRING)
+	: value_(std::make_shared<game_logic::variant_string>(str))
 {
-	string_ = new std::string(str);
+	assert(value_.get());
 }
 
 variant::variant(std::map<variant,variant>* map)
-    : type_(TYPE_MAP)
+	: value_((std::make_shared<game_logic::variant_map>(*map)))
 {
-	assert(map);
-	map_ = new std::map<variant,variant>(*map);
+	assert(value_.get());
 }
 
 variant::variant(const variant& v)
-    : type_(v.type_)
-{
-	switch(type_) {
-		case TYPE_INT:
-			int_value_ = v.int_value_;
-			break;
-		case TYPE_DECIMAL:
-			decimal_value_ = v.decimal_value_;
-			break;
-		case TYPE_LIST:
-			list_ = new std::vector<variant>(*v.list_);
-			break;
-		case TYPE_STRING:
-			string_ = new std::string(*v.string_);
-			break;
-		case TYPE_MAP:
-			map_ = new std::map<variant,variant>(*v.map_);
-			break;
-		case TYPE_CALLABLE:
-			callable_ = v.callable_;
-			break;
-		case TYPE_NULL:
-			break;
-	}
-}
-
-variant::~variant()
+    : value_(v.value_)
 {
-	release();
 }
 
 variant& variant::operator=(const variant& v)
 {
-	if(&v != this) {
-		this->~variant();
-		new(this) variant(v);
-	}
+	value_ = v.value_;
```

Comment:
If this is all it is, you can just remove the copy constructor altogether, because this is exactly what the default compiler-generated copy constructor would do.